File: client/src/app/models/upload.models.ts
export interface UploadInitiateRequest {
  fileName: string;
  fileSize: number;
  chunkSize?: number;
  contentType?: string;
  checksum?: string;
}

export interface UploadInitiateResponse {
  uploadId: string;
  fileName: string;
  fileSize: number;
  chunkSize: number;
  totalChunks: number;
  expiresAt: string;
}

export interface ChunkUploadRequest {
  uploadId: string;
  chunkNumber: number;
  checksum?: string;
}

export interface ChunkUploadResponse {
  uploadId: string;
  chunkNumber: number;
  success: boolean;
  isComplete: boolean;
  receivedChunks: number;
  totalChunks: number;
}

export interface UploadStatusResponse {
  uploadId: string;
  fileName: string;
  fileSize: number;
  state: UploadState;
  receivedChunks: number;
  totalChunks: number;
  missingChunks: number[];
  createdAt: string;
  updatedAt: string;
  expiresAt: string;
}

export interface UploadCompleteRequest {
  uploadId: string;
  finalChecksum?: string;
}

export interface UploadCompleteResponse {
  uploadId: string;
  fileName: string;
  fileSize: number;
  success: boolean;
  filePath?: string;
  errorMessage?: string;
  checksumVerified: boolean;
}

export enum UploadState {
  Initiated = 'Initiated',
  InProgress = 'InProgress',
  Completed = 'Completed',
  Failed = 'Failed',
  Expired = 'Expired'
}

export interface UploadProgress {
  uploadId: string;
  fileName: string;
  fileSize: number;
  progress: number;
  uploaded: number;
  status: 'pending' | 'uploading' | 'paused' | 'completed' | 'error';
  error?: string;
  speed?: number;
  remainingTime?: number;
}

export interface ChunkMetadata {
  index: number;
  start: number;
  end: number;
  size: number;
  uploaded: boolean;
  retries: number;
  checksum?: string;
}

export interface UploadSession {
  uploadId: string;
  file: File;
  chunks: ChunkMetadata[];
  totalChunks: number;
  chunkSize: number;
  uploadedChunks: number;
  progress: number;
  status: 'pending' | 'uploading' | 'paused' | 'completed' | 'error';
  error?: string;
  lastUpdated: Date;
  checksum?: string;
}

File: server/ResumableUpload.API/Models/UploadModels.cs
namespace ResumableUpload.API.Models;

public class UploadInitiateRequest
{
    public required string FileName { get; set; }
    public required long FileSize { get; set; }
    public int ChunkSize { get; set; } = 5 * 1024 * 1024; // Default: 5MB
    public string? ContentType { get; set; }
    public string? Checksum { get; set; }
}

public class UploadInitiateResponse
{
    public required string UploadId { get; set; }
    public required string FileName { get; set; }
    public long FileSize { get; set; }
    public int ChunkSize { get; set; }
    public int TotalChunks { get; set; }
    public DateTime ExpiresAt { get; set; }
}

public class ChunkUploadRequest
{
    public required string UploadId { get; set; }
    public required int ChunkNumber { get; set; }
    public string? Checksum { get; set; }
}

public class ChunkUploadResponse
{
    public required string UploadId { get; set; }
    public int ChunkNumber { get; set; }
    public bool Success { get; set; }
    public bool IsComplete { get; set; }
    public int ReceivedChunks { get; set; }
    public int TotalChunks { get; set; }
}

public class UploadStatusResponse
{
    public required string UploadId { get; set; }
    public required string FileName { get; set; }
    public long FileSize { get; set; }
    public UploadState State { get; set; }
    public int ReceivedChunks { get; set; }
    public int TotalChunks { get; set; }
    public List<int> MissingChunks { get; set; } = new List<int>();
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public DateTime ExpiresAt { get; set; }
}

public class UploadCompleteRequest
{
    public required string UploadId { get; set; }
    public string? FinalChecksum { get; set; }
}

public class UploadCompleteResponse
{
    public required string UploadId { get; set; }
    public required string FileName { get; set; }
    public long FileSize { get; set; }
    public bool Success { get; set; }
    public string? FilePath { get; set; }
    public string? ErrorMessage { get; set; }
    public bool ChecksumVerified { get; set; }
}

public enum UploadState
{
    Initiated,
    InProgress,
    Completed,
    Failed,
    Expired
}

public class UploadSession
{
    public required string UploadId { get; set; }
    public required string FileName { get; set; }
    public required long FileSize { get; set; }
    public int ChunkSize { get; set; }
    public int TotalChunks { get; set; }
    public UploadState State { get; set; } = UploadState.Initiated;
    public HashSet<int> ReceivedChunks { get; set; } = new HashSet<int>();
    public string? ContentType { get; set; }
    public string? Checksum { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    public DateTime ExpiresAt { get; set; }
}
